modules/3d/test/test_normal.cpp

```diff
@@ -396,12 +369,305 @@ class RgbdPlaneTest
+            {
+                Vec3f *nrmrow = srcNormals.ptr<Vec3f>(y);
+                for (int x = 0; x < srcNormals.cols; x++)
+                {
+                    Vec3f n = nrmrow[x];
+
+                    nrmrow[x] = rotm * n;
+                }
+            }
+        }
+        else if (depth == CV_64F)
+        {
+            // Inverted camera rotation
+            Matx33d rotm = rotm64;
+
+            for (int y = 0; y < srcNormals.rows; y++)
+            {
+                Vec3d *nrmrow = srcNormals.ptr<Vec3d>(y);
+                for (int x = 0; x < srcNormals.cols; x++)
+                {
+                    Vec3d n = nrmrow[x];
+
+                    nrmrow[x] = rotm * n;
+                }
+            }
+        }
+
+        Mat srcMask = srcDepth > 0;
+
+        float diffThreshold = 50.f;
+        if (scale)
+        {
+            srcDepth = srcDepth * 5000.0;
+            diffThreshold = 100000.f;
+        }
+
+        Mat srcCloud;
+        // The function with mask produces 1x(w*h) vector, this is not what we need
+        // depthTo3d(srcDepth, intr, srcCloud, srcMask);
+        depthTo3d(srcDepth, intr, srcCloud);
+        Scalar qnan = Scalar::all(std::numeric_limits<double>::quiet_NaN());
+        srcCloud.setTo(qnan, ~srcMask);
+        srcDepth.setTo(qnan, ~srcMask);
+
+        // For further result comparison
+        srcNormals.setTo(qnan, ~srcMask);
+
+        // TODO: test cases like above
+        Ptr<RgbdNormals> normalsComputer = RgbdNormals::create(1080, 1920, depth, intr, 5, diffThreshold, method);
+        normalsComputer->cache();
+
+        Mat dstNormals;
+        //TODO: LINEMOD can accept 16U and cloud instead of depth, fix it
+        if (normalsComputer->getMethod() == RgbdNormals::RGBD_NORMALS_METHOD_LINEMOD)
+        {
+            normalsComputer->apply(srcDepth, dstNormals);
+        }
+        else
+            normalsComputer->apply(srcCloud, dstNormals);
+
+        Result res;
+
+        // Remove 4th channel from dstNormals
+        Mat newDstNormals;
+        std::vector<Mat> dstNormalsCh;
+        split(dstNormals, dstNormalsCh);
+        dstNormalsCh.resize(3);
+        merge(dstNormalsCh, newDstNormals);
+        dstNormals = newDstNormals;
+
+        Mat dstMask(dstNormals.size(), CV_8U);
+        for (int y = 0; y < dstMask.rows; y++)
+        {
+            uchar* dstMaskRow = dstMask.ptr<uchar>(y);
+            if (depth == CV_32F)
+            {
+                Vec3f *dstrow = dstNormals.ptr<Vec3f>(y);
+                for (int x = 0; x < dstMask.cols; x++)
+                {
+                    dstMaskRow[x] = (dstrow[x] == dstrow[x])*255;
+                }
+            }
+            else if (depth == CV_64F)
+            {
+                Vec3d *dstrow = dstNormals.ptr<Vec3d>(y);
+                for (int x = 0; x < dstMask.cols; x++)
+                {
+                    dstMaskRow[x] = (dstrow[x] == dstrow[x])*255;
+                }
+            }
+        }
```

Comment:
Why not just use `compare`: https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga303cfb72acf8cbb36d884650c09a3a97 ?

Reply:
The purpose of this piece of code is to do NaN check; not sure if `compare` can be used here

Reply:
Moreover, `compare` performs comparison per channel, not per pixel